sonic_sku_create/sonic_sku_create.py

```diff
@@ -0,0 +1,329 @@
+
+Create a new SKU
+
+optional arguments:
+  -h, --help            show this help message and exit
+  -v, --version         show program's version number and exit
+  -f FILE, --file FILE  SKU definition file (default <cwd>/sku_def.xml)
+  -b BASE, --base BASE  SKU base definition
+  -r, --remove          Remove SKU folder
+  -c {new_sku_only,l2_mode_only,new_sku_l2}, --cmd {new_sku_only,l2_mode_only,new_sku_l2}
+                        Choose action to preform (Generate a new SKU, Configure L2 mode, Both
+  -l2_sku_name L2_SKU_NAME, --l2_sku_name L2_SKU_NAME
+                        SKU name to be used in the L2 configuration mode
+  -p, --print           Print port_config.ini without creating a new SKU
+  -vv, --verbose        Verbose output
+
+
+"""
+
+from __future__ import print_function
+
+import argparse
+import json
+import os
+import re
+import subprocess
+import swsssdk
+import sys
+import shutil
+
+from tabulate import tabulate
+import xml.etree.ElementTree as ET
+
+
+DEFAULT_DEV_PATH = '/usr/share/sonic/device/'
+### port_config.ini header
+PORTCONFIG_HEADER = ["# name", "lanes", "speed", "alias", "index"]
+
+
+class SkuCreate(object):
+	"""
+	Process aclstat
+	"""
+
+	def __init__(self):
+		self.metadata = {}
+		self.portconfig_dict = {}
+		self.platform_specific_dict = {"x86_64-mlnx_msn2700-r0":self.msn2700_specific}
+		self.default_lanes_per_port = []
+		self.platform = 'NA'
+		self.fpp = []
+		self.fpp_split  =  {}
+		self.num_of_fpp = 0
+		self.num_of_lanes_dict = {"1" : 4, "2" : 2, "4" : 1}
+		
+		# Set up db connections
+		self.db = swsssdk.SonicV2Connector(host="127.0.0.1")
+		self.db.connect(self.db.CONFIG_DB)
+		
+		self.configdb = swsssdk.ConfigDBConnector()
+		self.configdb.connect()
+
+		self.read_metadata()
+		self.platform = self.metadata['platform']
+		self.sku_name = None
+		self.base_sku_name = None
+		self.base_sku_dir = None
+		self.base_file_path = None		
+		self.new_sku_dir = None
+
+	def read_metadata(self):
+		"""
+		Redis database platform info
+		"""
+		self.metadata = self.db.get_all(self.db.CONFIG_DB, "DEVICE_METADATA|localhost" )
+
+	def sku_def_parser(self,sku_def,sku_base) :
+		try:
+			f = open(sku_def,"r")
+		except IOError:
+			print ("Couln't open file: " + sku_def)
+		element = ET.parse(f)
+		
+		root = element.getroot()
+		#print ( "tag=%s, attrib=%s" % (root.tag, root.attrib))
+		self.sku_name = root.attrib["HwSku"]
+		self.base_sku_name = sku_base
+		self.base_sku_dir = DEFAULT_DEV_PATH + self.platform + '/' + self.base_sku_name + '/'
+		self.base_file_path = self.base_sku_dir + "port_config.ini"
+		self.new_sku_dir = DEFAULT_DEV_PATH+self.platform+"/"+self.sku_name+ '/'
+		for child in root:
+			if child.tag == "Ethernet":
+				for interface in child:
+					for eth_iter in interface.iter():
+						if eth_iter is not None:
+							self.portconfig_dict[eth_iter.get("Index")] = ["Ethernet"+eth_iter.get("Index"),[1,2,3,4], eth_iter.get("Speed"), eth_iter.get("InterfaceName"),  eth_iter.get("Index")]
```

Comment:
>"Index" [](start = 41, length = 7)

I think we could ignore 'Index' in xml, and use InterfaceName as the clue.

Reply:
As a further design discussion, we could use InterfaceName in minigraph for new SKU to generate the SKU directory. If that is possible, could you help implement a new option in this script?

---
In reply to: [290541078](https://github.com/Azure/sonic-utilities/pull/547#discussion_r290541078) [](ancestors = 290541078)

Reply:
Why does it matter ?
Are you planning to remove the index from the minigraph ?

Reply:
You are assume the 'index' field is unique. If yes, can you remove the assumption. An internal 'index' based on row order is good enough, or remove it totally.

---
In reply to: [305300266](https://github.com/Azure/sonic-utilities/pull/547#discussion_r305300266) [](ancestors = 305300266)